tests/test_cache_engine.py

```diff
@@ -287,6 +287,85 @@ def test_skipping(fmt, autorelease):
+    chunk_size = 256
+    persist_path = "/tmp/test-engine-lookup.pth"
+
+    tokens = generate_tokens(num_tokens, device)
+    kv_cache = generate_kv_cache(num_tokens, fmt, device)
+    new_tokens = generate_tokens(new_num_tokens, device)
+    new_kv_cache = generate_kv_cache(new_num_tokens, fmt, device)
+    final_tokens = torch.cat([tokens, new_tokens])
+    final_kv_cache = concatenate_kv_caches([kv_cache, new_kv_cache], fmt)
+
+    cfg = LMCacheEngineConfig.from_legacy(chunk_size=chunk_size,
+                                          persist_path=persist_path)
+    engine = autorelease(LMCacheEngine(cfg, dumb_metadata(fmt)))
+
+    engine.store(tokens, kv_cache)
+
+    prefix_length = engine.lookup(tokens)
+    assert prefix_length == num_tokens, \
+        f"Expected {num_tokens} prefix tokens, but got {prefix_length}"
+
+    short_tokens_len = ((num_tokens // 2) // chunk_size) \
+        * chunk_size
+    short_tokens = tokens[:short_tokens_len]
+    prefix_length = engine.lookup(short_tokens)
+    assert prefix_length == short_tokens_len, \
+        f"Expected {short_tokens_len} prefix tokens, but got {prefix_length}"
+
+    prefix_length = engine.lookup(final_tokens)
+    expected_prefix_length = (num_tokens // chunk_size) * chunk_size
+    assert prefix_length == expected_prefix_length, \
+        f"Expected {expected_prefix_length} prefix tokens,"\
+            f" but got {prefix_length}"
+
+    engine.store(final_tokens, final_kv_cache)
+
+    final_prefix_length = engine.lookup(final_tokens)
+    assert final_prefix_length == num_tokens + new_num_tokens, \
+    f"Expected {num_tokens + new_num_tokens} prefix tokens,"\
+        f" but got {final_prefix_length}"
+
+
+@pytest.mark.parametrize("fmt", ["vllm", "huggingface"])
+def test_store_kv_tensors_mask(fmt, autorelease):
+    device = "cuda"
+    num_tokens = 12000
+    new_num_tokens = 2000
+    chunk_size = 256
+    persist_path = "/tmp/test-engine-store-kv-tensors-mask.pth"
+
+    tokens = generate_tokens(num_tokens, device)
+    kv_cache = generate_kv_cache(num_tokens, fmt, device)
+    new_tokens = generate_tokens(new_num_tokens, device)
+    final_tokens = torch.cat([tokens, new_tokens])
+
+    cfg = LMCacheEngineConfig.from_legacy(chunk_size=chunk_size,
+                                          persist_path=persist_path)
+    engine = autorelease(LMCacheEngine(cfg, dumb_metadata(fmt)))
+
+    engine.store(tokens, kv_cache)
+    prefix_length = engine.lookup(tokens)
+    assert prefix_length == num_tokens, \
+        f"Expected {num_tokens} prefix tokens, but got {prefix_length}"
+    kv_tensor_mask = torch.ones_like(final_tokens, dtype=torch.bool)
+    align_to_chunk_cnt = (num_tokens // chunk_size) * chunk_size
+    kv_tensor_mask[:align_to_chunk_cnt] = False
+    more_cache_tokens = num_tokens + new_num_tokens - align_to_chunk_cnt
+    more_kv_cache = generate_kv_cache(more_cache_tokens, fmt, device)
+    engine.store(final_tokens, more_kv_cache, kv_tensor_mask)
```

Comment:
Quick question: what should be the shape of `kv_tensors_raw` when calling `engine.store`? Should it have the same # of tokens as the tokens and `kv_tensors_mask`, or it should have the same # of tokens as the # of trues in the `kv_tensors_mask`?
My previous understanding is the former one, but here looks like `more_kv_cache` has less tokens than `final_tokens`

Reply:
The second. That's the way we reduce memory copy. If it is the first one, then no copy is reduced.

Reply:
Okay, maybe the latter one also makes sense. I realized only a "short" KV tensor will be extracted on the lmcache-vllm side. 

However, I feel like this could create some inconsistencies:
- If there is no `kv_tensors_mask`, the KV tensor and the tokens should be the same "length" (meaning the same number of tokens)
- If there is `kv_tensors_mask`, the KV tensor will have a different length from the input tokens.

In that case, I have another interface proposal: `store(prefix_tokens, new_tokens, KV_tensors)`, and the KV tensors have the same number of tokens in `new_tokens`. But maybe we can do that in the future.

@YaoJiayi Feel free to let us know if you have any thoughts on this

Reply:
> Quick question: what should be the shape of `kv_tensors_raw` when calling `engine.store`? Should it have the same # of tokens as the tokens and `kv_tensors_mask`, or it should have the same # of tokens as the # of trues in the `kv_tensors_mask`? My previous understanding is the former one, but here looks like `more_kv_cache` has less tokens than `final_tokens`

kv_tensors_mask means, if this position is true, then the corresponding token has its KV cache in kv_cache_raw passed into store.

Reply:
> The second. That's the way we reduce memory copy. If it is the first one, then no copy is reduced.

Can you add this to the docstring? Specifically, documenting the shape/device requirements explicitly will be useful.

Reply:
> > The second. That's the way we reduce memory copy. If it is the first one, then no copy is reduced.
> 
> Can you add this to the docstring? Specifically, documenting the shape/device requirements explicitly will be useful.

The docstring already has this:
```
            kv_tensors_mask: a boolean mask of tokens indicating which tokens'
            KV Cache should be stored. Only support suffix mask.
            len(kv_tensors_mask) should be the same as len(tokens)
            number of true should be the same as kv_tensors_raw token number
```
Do I need to write something else?

Reply:
Oh! I missed the last sentence, sorry about that!
Personally, I feel like it would be very useful to say something in the `kv_tensors_raw`'s description. For example "When kv_tensors_mask is None, the number of tokens in kv_tensors_raw should be the same as the number of Trues in kv_tensors_mask. Otherwise, it should be the same as len(tokens)".